### PR TITLE
CSS: Support recursive calc definition, improve compatibility for custom properties, support :has pseudo-class

### PR DESCRIPTION
**Pull request by matthiasblaesing**
**Support recursive CSS property definitions (calc function with embedded function calls)**

The grammar for calc property values failed to parse embedded var and
calc calls correctly. The issue is the work-around introduces as
function-content, which basicly prevents nesting any expression that
contains braces.

This changes the grammar parsing result from a tree to a directed graph,
which can contain cycles. This needs to taken into account by useers of
the parsed grammar.

An example of such a cycle is:

```
calc-fn -> calc-sum -> calc-product -> calc-value -> calc-sum
              ^                                        |
              +----------------------------------------+
```

For example this is now supported:

```css
.header {
  height: calc(28 * var(--spacing-unit));
}
```

**Improve compatibility for custom properties**

Custom properties can hold more or less arbitrary content. This was not parsed and is now:

```css
:root {
  --grid-gutter: (var(--margins) * var(--spacing-unit));
  --cell-margin: (var(--grid-gutter) / 2);
}
```

**Support CSS :has pseudo-class**

This was not parsed and is now:

```css
body:has(+ #aCheck:checked) {
    background-color:red;
}
```

Closes: #5025
Closes: #5303

### CODE REVIEW

File: ide/css.lib/src/org/netbeans/modules/css/lib/Css3.g [hunk 1]
```diff
@@ -36,6 +36,9 @@
 //    from the module directory (ide/css.lib)
 // 3. Rerun unittests
 // 4. Commit Css3.g together with generated Css3Lexer.java and Css3Parser.java
+//
```
Comment (vieiro) [inline ide/css.lib/src/org/netbeans/modules/css/lib/Css3.g:39]: At the top of this file there's a line like this:
```
Author : Jim Idle, Temporal Wave LLC.
```
We may want to add an additional author there (be it Matthias or the NetBeans team or something). After all we don't want Jim Idle to be blamed by our sins... ;-)

Comment (matthiasblaesing) [reply ide/css.lib/src/org/netbeans/modules/css/lib/Css3.g:39]: I'm inclined to agree, however I quote:

```
// This grammar is free to use providing you retain everyhting in this header comment
// section.
```

so I let it stand as is. Lines 19+ I think spell it out, that this is a modified version.

File: ide/css.lib/src/org/netbeans/modules/css/lib/api/properties/Properties.java [hunk 2]
```diff
@@ -144,24 +146,26 @@ public static boolean isAggregatedProperty(FileObject context, PropertyDefinitio
         final GroupGrammarElement grammarElement = propertyDefinition.getGrammarElement(context);
         final AtomicBoolean isAggregated = new AtomicBoolean(false);
         grammarElement.accept(new GrammarElementVisitor() {
-            private boolean cancelled = false;
+            private final Map<GroupGrammarElement,Void> seen = new IdentityHashMap<>();
```
Comment (vieiro) [inline ide/css.lib/src/org/netbeans/modules/css/lib/api/properties/Properties.java:149]: It's ok. Maybe`Collections.newSetFromMap(new IdentityHashMap<>()) `  makes things clearer below (`seen.contains`, `seen.add`, etc.)?

File: ide/css.lib/src/org/netbeans/modules/css/lib/properties/GrammarParser.java [hunk 3]
```diff
@@ -372,4 +386,59 @@ public CharSequence readText() {
...
+            return truncatedPath(this.parent).equals(truncatedPath(other.parent));
+        }
+
+        @SuppressWarnings("AssignmentToMethodParameter")
+        private String truncatedPath(GrammarElement parent) {
+            StringBuilder path = new StringBuilder();
+            while(parent != null) {
+                path.append(parent.getName());
+                path.append("/");
+                if(parent.getName() != null) {
```
Comment (vieiro) [inline ide/css.lib/src/org/netbeans/modules/css/lib/properties/GrammarParser.java:436]: This `if (parent.getName() != null) {` shouldn't it be two lines above? Or do we want `path.append(parent.getName());` even if `parent.getName()` is null?

File: ide/css.visual/src/org/netbeans/modules/css/visual/RuleEditorNode.java [hunk 4]
```diff
@@ -595,15 +596,29 @@ private PropertyValuesEditor createPropertyValueEditor(FileObject context, Prope
             GroupGrammarElement rootElement = pmodel.getGrammarElement(context);
 
             rootElement.accept(new GrammarElementVisitor() {
+                private Map<GroupGrammarElement,Void> seen = new IdentityHashMap<>();
+
```
Comment (vieiro) [inline ide/css.visual/src/org/netbeans/modules/css/visual/RuleEditorNode.java:600]: It's fine, as in comment above we could also use a map for cleaner semantics.

**Review by vieiro — APPROVED** [review]
Great job! Thanks, @matthiasblaesing

File: ide/css.visual/src/org/netbeans/modules/css/visual/RuleEditorNode.java (hunk 4 above)
Comment (matthiasblaesing) [reply ide/css.visual/src/org/netbeans/modules/css/visual/RuleEditorNode.java:601]: Addressed via https://github.com/apache/netbeans/pull/5432/commits/e2dd93481e809985df55c138448cbaaf0ab75f4b

File: ide/css.lib/src/org/netbeans/modules/css/lib/api/properties/Properties.java (hunk 2 above)
Comment (matthiasblaesing) [reply ide/css.lib/src/org/netbeans/modules/css/lib/api/properties/Properties.java]: Addressed via https://github.com/apache/netbeans/pull/5432/commits/e2dd93481e809985df55c138448cbaaf0ab75f4b
Learned something new here - thanks.

File: ide/css.lib/src/org/netbeans/modules/css/lib/properties/GrammarParser.java (hunk 3 above)
Comment (matthiasblaesing) [reply ide/css.lib/src/org/netbeans/modules/css/lib/properties/GrammarParser.java]: Yes implementation of `PropertyGrammarElementRef` was bogus. Addressed via:
https://github.com/apache/netbeans/pull/5432/commits/ecd64eee0b30d600967c751cae5281cf842daa1a

**Comment by matthiasblaesing** [comment]
@vieiro thank you for the review. I pushed updates - could you have another look?

**Comment by vieiro** [comment]
> @vieiro thank you for the review. I pushed updates - could you have another look?

@matthiasblaesing Great job. Thanks again for this!

**Pull request merged**